### PR TITLE
BitDecomposition in eclair

### PR DESCRIPTION
**Pull request by SupremoUGH**
Bit decomposition in eclair:

Changes: 
`eclair`: 
- [x] `bool.rs`: Defined the `BitDecomposition` trait, implemented it for the native compiler and the usual types.
- [x] `num.rs`: Impl for `UnsignedInteger`
- [x] define `to_bits_be()` in the trait with default impl

`plugins::arkworks`
- [x] `constraint`: Impl for `FpVar`
- [x] `algebra`: Impl for `ScalarVar` 

---

Before we can merge this PR, please make sure that all the following items have been checked off:

- [x] Linked to an issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** line describing your change in [`CHANGELOG.md`](https://github.com/openzklib/openzl/blob/main/CHANGELOG.md) and added the appropriate `L-` label to the PR.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [x] Checked that changes and commits conform to the standards outlined in [`CONTRIBUTING.md`](https://github.com/openzklib/openzl/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: plugins/arkworks/src/constraint.rs [hunk 1]
```diff
@@ -261,6 +271,22 @@ where
...
+    fn to_bits_le(&self, compiler: &mut R1CS<F>) -> [Boolean<F>; BITS] {
+        let _ = compiler;
+        assert_eq!(
+            BITS,
+            F::Params::MODULUS_BITS as usize,
+            "BITS must be equal to MODULUS BITS"
+        );
+        ToBitsGadget::to_bits_le(self).expect("Bit decomposition is not allowed to fail.").try_into().expect("Obtaining an array of size BITS from a vector of length BITS is not allowed to fail.")
+    }
+}
```
Comment (bhgomes) [inline plugins/arkworks/src/constraint.rs:288]: we can't use `F::Params::MODULUS_BITS` here for `BitDecomposition<X, R1CS<F>>`?

File: plugins/arkworks/src/constraint.rs [hunk 2]
```diff
@@ -261,6 +271,22 @@ where
...
+{
+    #[inline]
+    fn to_bits_le(&self, compiler: &mut R1CS<F>) -> [Boolean<F>; BITS] {
+        let _ = compiler;
+        assert_eq!(
+            BITS,
+            F::Params::MODULUS_BITS as usize,
+            "BITS must be equal to MODULUS BITS"
+        );
+        ToBitsGadget::to_bits_le(self).expect("Bit decomposition is not allowed to fail.").try_into().expect("Obtaining an array of size BITS from a vector of length BITS is not allowed to fail.")
```
Comment (bhgomes) [inline plugins/arkworks/src/constraint.rs:286]: This always has trailing zeroes?

File: plugins/arkworks/src/constraint.rs [hunk 3]
```diff
@@ -124,9 +124,8 @@ where
             BITS < F::Params::MODULUS_BITS as usize,
             "BITS must be strictly less than modulus bits of `F`."
         );
-        let value_bits = value
-            .to_bits_le()
-            .expect("Bit decomposition is not allowed to fail.");
+        let value_bits =
+            ToBitsGadget::to_bits_le(value).expect("Bit decomposition is not allowed to fail.");
```
Comment (bhgomes) [inline plugins/arkworks/src/constraint.rs:128]: Should we use bit decomposition trait here and use this to build a generic impl of `assert_within_range`. If you think it's a good idea mark an issue for it.

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: plugins/arkworks/src/constraint.rs (hunk 1 above)
Comment (SupremoUGH) [reply plugins/arkworks/src/constraint.rs]: Why is that?

File: plugins/arkworks/src/constraint.rs (hunk 2 above)
Comment (SupremoUGH) [reply plugins/arkworks/src/constraint.rs]: just tested it. It does.

File: plugins/arkworks/src/constraint.rs (hunk 3 above)
Comment (SupremoUGH) [reply plugins/arkworks/src/constraint.rs]: https://github.com/openzklib/openzl/issues/55

File: plugins/arkworks/src/constraint.rs (hunk 1 above)
Comment (SupremoUGH) [reply plugins/arkworks/src/constraint.rs]: rust doesn't let me do the clean implementation 
impl<F> BitDecomposition<F::Params::MODULUS_BITS, R1CS<F>> for FpVar<F>

File: plugins/arkworks/src/constraint/mod.rs [hunk 4]
```diff
@@ -569,4 +596,32 @@ mod tests {
...
+            .collect::<Vec<_>>();
+        assert_eq!(
+            bit_decomposition_le,
+            bit_decomposition_be
+                .clone()
+                .into_iter()
+                .rev()
+                .collect::<Vec<_>>(),
+            "Little-endian and big-endian representations of number are not each other's reverse."
+        );
```
Comment (GhostOfGauss) [inline plugins/arkworks/src/constraint/mod.rs:623]: This test would be more convincing if `number` were randomly sampled and we compared to its bit decomposition as a u8 (or u64 or something)

**Review by GhostOfGauss — COMMENTED** [review]
Looks good I just suggest improving the correctness test

File: plugins/arkworks/src/constraint/mod.rs (hunk 4 above)
Comment (bhgomes) [reply plugins/arkworks/src/constraint/mod.rs:623]: Move the comparison into a function and then call it on small numbers 0, 1, 2, and then a random `Fr` point.

**Review by GhostOfGauss — APPROVED** [review]

**Review by bhgomes — APPROVED** [review]

**Pull request merged**